src/net/delta.cpp

```diff
@@ -0,0 +1,139 @@
+    return getSerImpl(getSerPtr(obj), make_index_sequence<tuple_size<decltype(getSerPtr(obj))>::value>());
+}
+
+/* helper functions */
+
+using expander = int[]; // https://stackoverflow.com/a/25683817
+
+template <size_t I, typename ... Args>
+void diffHeaderIter(const tuple<Args...> &left, const tuple<Args...> &right, uint16_t &header) {
+    if (get<I>(left) != get<I>(right)) {
+        header |= 1 << I;
+    }
+}
+
+template <size_t ... Is, typename ... Args>
+uint16_t diffHeader(const tuple<Args...> &left, const tuple<Args...> &right, integer_sequence<size_t, Is...>) {
+    uint16_t header = 0;
+    expander{ 0, ((void)diffHeaderIter<Is>(left, right, header), 0) ... }; // pseudo fold-expansion
+    return header;
+}
+
+template <size_t I, typename ... Args>
+void writeDeltaIter(StreamWriter &writer, const tuple<Args...> &o, uint16_t &header) {
+    if (header  &(1 << I)) {
+        writer << get<I>(o);
+    }
+}
+
+template <size_t ... Is, typename ... Args>
+void writeDeltaImpl(StreamWriter &writer, const tuple<Args...> &o, uint16_t &header, integer_sequence<size_t, Is...>) {
+    expander{ 0, ((void)writeDeltaIter<Is>(writer, o, header), 0) ... }; // pseudo fold-expansion
+}
+
+
+template <size_t I, typename ... Args>
+void readDeltaIter(StreamReader &reader, tuple<Args*...> &optr, uint16_t &header) {
+    if (header  &(1 << I)) {
+        reader >> *(get<I>(optr));
+    }
+}
+
+template <size_t ... Is, typename ... Args>
+void readDeltaImpl(StreamReader &reader, tuple<Args*...> &&optr, uint16_t &header, integer_sequence<size_t, Is...>) {
+    expander{ 0, ((void)readDeltaIter<Is>(reader, optr, header), 0) ... };
+}
+
+/* END helper functions */
+
+template <typename ... Args>
+void writeDeltaBuff(const tuple<Args...>&cur, const tuple<Args...> &old, StreamWriter &writer) {
+    uint16_t header = diffHeader(cur, old, make_index_sequence<sizeof...(Args)>());
+    writer << header;
+    writeDeltaImpl(writer, cur, header, make_index_sequence<sizeof...(Args)>());
+}
+
+template <typename T>
+void writeDelta(StreamWriter &writer, T &cur, T &old) {
+    writeDeltaBuff(getSerial(cur), getSerial(old), writer);
+}
+
+/* r-value to discourage outside scope usage*/
+template <typename ... Args>
+void readDeltaBuff(tuple<Args*...> &&obj, StreamReader &reader) {
+    uint16_t header = 0;
+    reader >> header;
+    readDeltaImpl(reader, move(obj), header, make_index_sequence<sizeof...(Args)>());
+}
+
+template <typename T>
+void readDelta(StreamReader &reader, T &obj) {
+    readDeltaBuff(getSerPtr(obj), reader);
+}
+
+// https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl
+#define DECL_SERIAL(T, ...) auto getSerPtr(T &obj) \
+    { return make_tuple(__VA_ARGS__); } \
+    template void writeDelta<T>(StreamWriter &writer, T &cur, T &old); \
+    template void readDelta<T>(StreamReader &reader, T &obj);
+
+DECL_SERIAL(CreatureStatus, &obj.x, &obj.y, &obj.z, &obj.heading, &obj.anim, &obj.animframe, &obj.faction);
```

Comment:
libnet is lower level than libmp, so it shouldn't have to depend on `CreatureStatus`.